galley/pkg/config/processing/runtime.go

```diff
@@ -0,0 +1,167 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package processing
+
+import (
+	"sync"
+	"sync/atomic"
+
+	"istio.io/api/mesh/v1alpha1"
+
+	"istio.io/istio/galley/pkg/config/event"
+)
+
+// ProcessorOptions are options that are passed to event.Processors during startup.
+type ProcessorOptions struct {
+	MeshConfig   *v1alpha1.MeshConfig
+	DomainSuffix string
+}
+
+// RuntimeOptions is options for Runtime
+type RuntimeOptions struct {
+	Sources      []event.Source
+	Processor    event.Processor
+	DomainSuffix string
+}
+
+// Clone returns a cloned copy of the RuntimeOptions.
+func (o *RuntimeOptions) Clone() RuntimeOptions {
+	sources := make([]event.Source, len(o.Sources))
+	copy(sources, o.Sources)
+
+	return RuntimeOptions{
+		Sources:      sources,
+		Processor:    o.Processor,
+		DomainSuffix: o.DomainSuffix,
+	}
+}
+
+// Runtime is the config processing runtime.
+type Runtime struct { // nolint:maligned
```

Comment:
I'm not sure I fully understand the distinction between runtime and session. After a quick glance, I can't determine why runtime is needed at all.

Reply:
Ok so it sounds like the main purpose of the class is to handle `RESET` events and mesh config updates.

Can you clarify the use case for a `RESET` event? When would it happen? Do we have something similar today?

Reply:
Thanks that helps ... so RESET events will be relatively infrequent by the sound of it.

Reply:
I'm still struggling a bit with the naming ... runtime vs session.

Is the `runtime` effectively just a node in the processing pipeline? In other words, the runtime is a hard-wired thing in the pipeline whose internal state may be restarted due to various events.

If this is the conceptual model, maybe renaming them to `node` and `state` would help to clarify?

Reply:
What do you mean by `the whole thing`? There seems to be a one-to-one mapping between `Runtime` and `Session`, and `Session` to `Processor`. If that's true then a `Runtime` maps one-to-one with a `Processor` ... so which would imply that it's just a singe node within a greater processing pipeline ... or am I missing something?